### PR TITLE
[go function] support localrun and cluster mode for go function

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <ranxiaolong716@gmail.com>

### Motivation

Master Issue: #3767 

support  local-run and cluster mode for go function.

in go function, we can use:

```
./bin/pulsar-admin functions localrun/create  
--go /Users/wolf4j/github.com/apache/pulsar/pulsar-function-go/examples/outputFunc.go 
--inputs persistent://public/default/my-topic 
--output persistent://public/default/test 
--tenant public 
--namespace default 
--name pulsarfunction 
--classname hellopulsar 
--log-topic logtopic
```

Different from `--jar` or `--py`, `--go` uploads a complete executable file(including: instance file + user code file)

### CODE REVIEW

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.functions.instance.functionforgo;
```
Comment (sijie) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java]: How about just calling this package "org.apache.pulsar.functions.instance.go"? That means it is the package related to go instance.

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import lombok.Getter;
+import lombok.Setter;
+
+@Setter
+@Getter
+public class FunctionConverGoConfig {
```
Comment (sijie) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java]: How about calling it "GoInstanceConfig"?

File: pulsar-functions/runtime/pom.xml [hunk 3]
```diff
@@ -50,6 +50,13 @@
       <artifactId>jcommander</artifactId>
     </dependency>
 
+    <dependency>
```
Comment (sijie) [inline pulsar-functions/runtime/pom.xml]: Can you define `snakeyaml` in the root pom file? So that you can just reference the artifact without specifying the version. Pulsar manages all the versions in the root pom file.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java [hunk 4]
```diff
@@ -190,6 +190,8 @@
             case PYTHON:
                 logConfigFile = pulsarRootDir + "/conf/functions-logging/console_logging_config.ini";
                 break;
+            case GO:
+                break;
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntime.java]: is this supported? if it is not fully supported, I would suggest throwing UnsupportedOperationException in this place until we confirm it is working.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntimeFactory.java [hunk 5]
```diff
@@ -185,14 +185,16 @@ public boolean externallyManaged() {
...
-        String instanceFile;
+        String instanceFile = null;
         switch (instanceConfig.getFunctionDetails().getRuntime()) {
             case JAVA:
                 instanceFile = javaInstanceJarFile;
                 break;
             case PYTHON:
                 instanceFile = pythonInstanceFile;
                 break;
+            case GO:
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntimeFactory.java]: is this supported? if it is not fully supported, I would suggest throwing UnsupportedOperationException in this place until we confirm it is working.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 6]
```diff
@@ -89,31 +91,180 @@
...
+     * When we run the go function, we only need to specify the location of the go-function file and the yaml file.
+     * The content of the yaml file will be automatically generated according to the content provided by instanceConfig.
+     *
+     */
+
+    public static List<String> getGoInstanceCmd(InstanceConfig instanceConfig,
+                                                String originalCodeFileName,
+                                                String pulsarServiceUrl) {
+        final List<String> args = new LinkedList<>();
+        FunctionConverGoConfig functionConverGoConfig = new FunctionConverGoConfig();
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: How about renaming it to "goInstanceConfig"? That sounds cleaner.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 7]
```diff
@@ -89,31 +91,180 @@
...
+        }
+
+        functionConverGoConfig.setKillAfterIdleMs(0);
+
+        DumperOptions dumperOptions = new DumperOptions();
+        dumperOptions.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+        Yaml yaml = new Yaml(dumperOptions);
+        String output = yaml.dumpAs(functionConverGoConfig, Tag.MAP, null);
+        String pathName = "pulsar-functions/runtime/src/main/resources";
+        String fileName = String.format("%s_%s_%s.yml", functionConverGoConfig.getTenant(), functionConverGoConfig.getNameSpace(),
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: I would suggest using TempFile for now. So the config file is deleted when the instance dies so we don't need to manage those configuration files.

```
Path path = Files.createTempFile(...);
```

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 8]
```diff
@@ -89,31 +91,180 @@
...
+                functionConverGoConfig.getName());
+        File ymlFile = new File(pathName + "/" + fileName);
+        try {
+            FileWriter fileWriter = new FileWriter(ymlFile);
+            fileWriter.write(output);
+            fileWriter.flush();
+            fileWriter.close();
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: it is a bad practice to use `e.printStackTrace()`. 

Also I don't think you should catch the exception here. You should let the function throw the exception and fail the operation. Because Go instance can not run correctly without the instance config.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 9]
```diff
@@ -89,31 +91,180 @@
...
         }
 
         return args;
     }
 
+    /**
+     *
+     * Different from python and java function, Go function uploads a complete executable file(including:
+     * instance file + user code file). Its parameter list is provided to the broker in the form of a yaml file,
+     * the advantage of this approach is that backward compatibility is guaranteed.
```
Comment (sijie) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: I would suggest expanding more details on what does the backward compatibility mean here.

If I understand this correctly, in Java and Python the instance is managed by broker (or function worker) so the changes in command line is under control; but in Go the instance is compiled with the user function, so pulsar doesn't have the control what instance is used in the function. Hence in order to support BC for go function, we can't dynamically add more commandline arguments. Using an instance config to pass the parameters from function worker to go instance is the best way for maintaining the BC. 

Am I understanding this correctly?

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/DefaultSecretsProviderConfigurator.java [hunk 10]
```diff
@@ -40,6 +40,8 @@ public String getSecretsProviderClassName(Function.FunctionDetails functionDetai
                 return ClearTextSecretsProvider.class.getName();
             case PYTHON:
                 return "secretsprovider.ClearTextSecretsProvider";
+            case GO:
```
Comment (sijie) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/DefaultSecretsProviderConfigurator.java]: throw new UnsupportedOperationException if a feature is not implemented.

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java [hunk 11]
```diff
@@ -47,6 +47,8 @@ public String getSecretsProviderClassName(Function.FunctionDetails functionDetai
                 return EnvironmentBasedSecretsProvider.class.getName();
             case PYTHON:
                 return "secretsprovider.EnvironmentBasedSecretsProvider";
+            case GO:
+                return "";
```
Comment (sijie) [inline pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/KubernetesSecretsProviderConfigurator.java]: throw new UnsupportedOperationException if a feature is not implemented.

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 12]
```diff
@@ -429,6 +433,20 @@ private static void doPythonChecks(FunctionConfig functionConfig) {
         }
     }
 
+    private static void doGolangChecks(FunctionConfig functionConfig) {
+        if (functionConfig.getProcessingGuarantees() == FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE) {
+            throw new RuntimeException("Effectively-once processing guarantees not yet supported in Golang");
+        }
+
+        if (functionConfig.getWindowConfig() != null) {
+            throw new IllegalArgumentException("There is currently no support windowing in golang");
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java]:
```suggestion
            throw new IllegalArgumentException("Windowing is not supported in Go function yet");
```

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 13]
```diff
@@ -429,6 +433,20 @@ private static void doPythonChecks(FunctionConfig functionConfig) {
         }
     }
 
+    private static void doGolangChecks(FunctionConfig functionConfig) {
+        if (functionConfig.getProcessingGuarantees() == FunctionConfig.ProcessingGuarantees.EFFECTIVELY_ONCE) {
+            throw new RuntimeException("Effectively-once processing guarantees not yet supported in Golang");
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java]:
```suggestion
            throw new RuntimeException("Effectively-once processing guarantees not yet supported in Go function");
```

File: pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java [hunk 14]
```diff
@@ -429,6 +433,20 @@ private static void doPythonChecks(FunctionConfig functionConfig) {
...
+        }
+
+        if (functionConfig.getWindowConfig() != null) {
+            throw new IllegalArgumentException("There is currently no support windowing in golang");
+        }
+
+        if (functionConfig.getMaxMessageRetries() != null && functionConfig.getMaxMessageRetries() >= 0) {
+            throw new IllegalArgumentException("Message retries not yet supported in golang");
```
Comment (sijie) [inline pulsar-functions/utils/src/main/java/org/apache/pulsar/functions/utils/FunctionConfigUtils.java]:
```suggestion
            throw new IllegalArgumentException("Message retries not yet supported in Go function");
```

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java [hunk 15]
```diff
@@ -501,13 +511,13 @@ protected void validateFunctionConfigs(FunctionConfig functionConfig) {
                 org.apache.pulsar.common.functions.Utils.inferMissingNamespace(functionConfig);
             }
 
-            if (isNotBlank(functionConfig.getJar()) && isNotBlank(functionConfig.getPy())) {
-                throw new ParameterException("Either a Java jar or a Python file needs to"
+            if (isNotBlank(functionConfig.getJar()) && isNotBlank(functionConfig.getPy()) && isNotBlank(functionConfig.getGo())) {
+                throw new ParameterException("Either a Java jar or a Python file or a Go file needs to"
```
[the remark this answers is not in the document]
Comment (wolfstudy) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: yes, in here,  we uploaded a source files suffixed with `.go`.

in [getGoInstanceCmd()](https://github.com/apache/pulsar/pull/4174/files#diff-092ae104654a1e712a083dacc1ee5308R233) 

```
args.add("go");
args.add("run");
args.add(originalCodeFileName);
```

We use go run to run this go file, `go run` can be divided into two steps:

1. build `.go` file is an executable binary
2. run this executable binary

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java (hunk 1 above)
Comment (wolfstudy) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java]: Excellent, `go` is a better choice than `functionforgo`

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java (hunk 2 above)
Comment (wolfstudy) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/functionforgo/FunctionConverGoConfig.java]: cool, will fix it .

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (wolfstudy) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: yes, completely correct,

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sijie) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: @wolfstudy 

Actually I forgot adding the comment there. I don't think we should use "go run", because not every machine installed `go`. so I would prefer uploading the executable binary.

[the remark this answers is not in the document]
Comment (wolfstudy) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdFunctions.java]: awesome, will fix it

File: pulsar-functions/runtime/pom.xml (hunk 3 above)
Comment (jerrypeng) [reply pulsar-functions/runtime/pom.xml]: @wolfstudy can we not use Jackson which is already dependency to write the yaml file?

File: pom.xml [hunk 16]
```diff
@@ -450,6 +450,12 @@ flexible messaging model and an intuitive client API.</description>
         <version>1.48</version>
       </dependency>
 
+      <dependency>
```
Comment (jerrypeng) [inline pom.xml]: Same comment about using Jackson instead

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 17]
```diff
@@ -21,9 +21,7 @@
 
 import com.google.protobuf.util.JsonFormat;
 
-import java.io.BufferedReader;
-import java.io.IOException;
-import java.io.InputStreamReader;
+import java.io.*;
```
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: Please don't use wildcards

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: @wolfstudy I don't understand how this helps maintain BC.  It just seems like we are pushing to BC problem to somewhere else.  If we make a change to instance config, wouldn't BC problem still occur?

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/LocalRunner.java [hunk 18]
```diff
@@ -108,6 +108,8 @@ void start() throws Exception {
                     }
                     classLoader = loadJar(file);
                 }
+            } else if (functionConfig.getRuntime() == FunctionConfig.Runtime.GO) {
+                userCodeFile = functionConfig.getGo();
```
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/LocalRunner.java]: Can we also check for python runtime here and throw an exception if its not Java, Go, or python

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: if you shipped a go instance with `--flag1` and `--flag2`. the instance is compiled with user code. If a new broker introduce a new flag for running instance `--flag3`. The old go function (instance compiled with user code) doesn't know `--flag3`. so this instance can not run with new broker code. so you can't upgrade the broker.

in what @wolfstudy 's approach, the parameters is not passing from command lines from broker (function worker) to instance. instead it is using an instance conf file. so there is no need to change command line arguments for go instance. It becomes easier to add fields to a yaml file, because ignoring new (unknown) fields in yaml is much easier.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 19]
```diff
@@ -82,38 +84,183 @@
...
+
+        goInstanceConfig.setKillAfterIdleMs(0);
+
+        DumperOptions dumperOptions = new DumperOptions();
+        dumperOptions.setDefaultFlowStyle(DumperOptions.FlowStyle.BLOCK);
+        Yaml yaml = new Yaml(dumperOptions);
+        String output = yaml.dumpAs(goInstanceConfig, Tag.MAP, null);
+        String fileName = String.format("%s_%s_%s", goInstanceConfig.getTenant(), goInstanceConfig.getNameSpace(),
+                goInstanceConfig.getName());
+        File ymlFile = File.createTempFile(fileName, ".yml");
```
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: This file needs to be cleaned somehow.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > It becomes easier to add fields to a yaml file, because ignoring new (unknown) fields in yaml is much easier.

I not a GO expert but is there not a flag for the arg parsers in GO to ignore unknown fields?  

The downside to passing in a file is that you have to clean it up somehow.  Even temporary files need to be cleaned up or they will start to pile up.  Also this creates difficulties for container based runtimes like K8s.  How shall we pass the file to a container?  While its not impossible it does add complications.

Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > I not a GO expert but is there not a flag for the arg parsers in GO to ignore unknown fields?

Can't answer that since I am not a Go export.

But I doubt there is a such thing in arg parsers in any language. Because as an args parser, you need to distinguish what are options and what are args. without a predefined list of options, it is hard to know what are options and what are args, hence it is hard to know what can be ignored.

>  you have to clean it up somehow.

In process mode, you can do File.deleteOnExit. Because the file is only used during execution time. You don't really need to save it. 

> Also this creates difficulties for container based runtimes like K8s.  How shall we pass the file to a container? 

1. the yaml file can be save to a configmap.
2. container can mount the configmap as the file
3. parameters that depends on instance (e.g. instance id) can be passed as environment variables. parameters can be substituted by running the apply-config script similar to what we do for downloading the codeFile.


> While its not impossible it does add complications.

Every language has its own *complication*. In java, you have to handle classpath and shading; in Python, you have to handle dependencies, pywheel, and zip file; In Go, you have to handle the situations introduced due to Go is a static-linking language. IMO this is not "complication". They are just different situations introduced by different languages.

A better approach here (in future) is to have a language abstraction. Broker / Function worker only passes the function config and necessary metadata to the language abstraction. Each language chooses its best way to construct the command to invoke the function. So the language specific details can be hidden from broker / function worker. Broker / Function worker doesn't care if a language is passing the function config in a file to the instance or passing the function config through command args.

Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > In process mode, you can do File.deleteOnExit. Because the file is only used during execution time. You don't really need to save it.

``` File.deleteOnExit ``` deletes the file when the process that created it exists, but the process that created this file is not the function instance process, its the worker process.  The worker process can run forever.

> But I doubt there is a such thing in arg parsers in any language. Because as an args parser, you need to distinguish what are options and what are args. without a predefined list of options, it is hard to know what are options and what are args, hence it is hard to know what can be ignored.

In python you can do the following:

```
args, unknown = parser.parse_known_args()
```

I get that this would be impossible fo positional arguments, but we are not using any positional arguments.  An parameters passed to the instance is in the form of ```--argument_key <augument_value>```

looking at how arguments are parsed in GO:

https://gobyexample.com/command-line-flags

It seems to me that you explicitly specify which arguments to parse and the ones you don't just get ignored.  Which will be fine for BC as long as we don't remove any arguments.

> Every language has its own complication. In java, you have to handle classpath and shading; in Python, you have to handle dependencies, pywheel, and zip file; In Go, you have to handle the situations introduced due to Go is a static-linking language. IMO this is not "complication". They are just different situations introduced by different languages.

Sure we might have to do different things for different languages, but also it is better if we can minimize these differences so that we don't have to write a completely different code path to handle each language if possible.

If we can start the GO function without using a file, integration with k8 will require little/no work

Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > File.deleteOnExit deletes the file when the process that created it exits, but the process that created this file is not the function instance process, its the worker process. The worker process can run forever.

Correct. You need to explicitly delete the file if the instance is stopped due to a function is stopped or deleted.

> It seems to me that you explicitly specify which arguments to parse and the ones you don't just get ignored. 

I don't think so. That's what I have tried.

```
$ go run test.go -wat
flag provided but not defined: -wat
Usage of /var/folders/_l/j6x4h5m90l538hpmdj55l2f80000gn/T/go-build919491998/b001/exe/test:
  -fork
    	a bool
  -numb int
    	an int (default 42)
  -svar string
    	a string var (default "bar")
  -word string
    	a string (default "foo")
exit status 2
```

 > I get that this would be impossible for positional arguments, but we are not using any positional arguments. An parameters passed to the instance is in the form of --argument_key <augument_value>

When talking about BC, it is not about what current is. It is about what the future will be.  We are not using positional arguments for now. but will it be used in future?

> but also it is better if we can minimize these differences so that we don't have to write a completely different code path to handle each language if possible.

Agreed with minimizing the differences. Though I think minimizing the differences should be done in a way that makes sense to all languages. Currently I don't think the command-line argument approach work well for go function in general.

Thinking about the question in a different way -  Is the command-line arg approach the best option for all languages? Does using an instance conf for passing the function config and option work better for all languages?

At least, it was not a fun to me when debugging any instance related issues for Java and Python. Because a) the command line is just too long; b) you have to get all function details encoded to a string that is able to be passed in a command line; c) the encoded function details is almost not human-readable. 

Hence if Go has to use the file approach due to addressing BC concerns, for the sake of uniformity, does it make sense (or is it better) to standardize in the file approach. I don't expect us to do any changes in this pull request or soon, but I DO expect people to think about what is the best approach in long term, not just stick to what currently we have.

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 17 above)
Comment (wolfstudy) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: cool, good suggestion

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: I think there is a config you can set to to ignore unknown flags. Please reference https://godoc.org/github.com/jessevdk/go-flags.

```
Ignoring unknown command line options (optional)
```

Even if are using files to passing arguments to the GO instance why does it have to be the arguments in instance config?  Why not just write the command line arguments in a file to begin with?  Given we can ignore unknown flags in GO, then BC i.e. adding new flags shouldn't be a problem.

> When talking about BC, it is not about what current is. It is about what the future will be. We are not using positional arguments for now. but will it be used in future?

I don't think we will ever add positional arguments because that will become a BC issue for all the runtimes.

> At least, it was not a fun to me when debugging any instance related issues for Java and Python. Because a) the command line is just too long; b) you have to get all function details encoded to a string that is able to be passed in a command line; c) the encoded function details is almost not human-readable.

These are superficial problems right. A user can also copy and paste the command line into a text editor and then parse it however he/she would like.  

However, using a file to pass arguments to function instances will create fundamental challenges we will have to solve.  Sure for K8s, we can maybe use a configmap and mount it to pods, but in the future what about other runtimes.  For example, if we wanted to support just a docker runtime how would we do that.  We would maybe need to mount a directory from local disk to the container.  What about if we want to support another external scheduler like Nomad or Mesos, it has it own way of mapping files into containers.  My point is adding this file dependency will create more moving parts and leave more room for error and, as far as, I can tell every scheduler allows your to specify command line arguments to run a process or container but placing files to where your container/process will run can be widely different.  On top of that we will also have to figure out how to clean up these files for non-container based runtimes.

While all of this is not impossible to do, my question is it really worth it?

**Comment by wolfstudy** [comment]
@jerrypeng PTAL again thanks

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java (hunk 9 above)
Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > Even if we are using files to passing arguments to the GO instance why does it have to be the arguments in instance config? Why not just write the command line arguments in a file to begin with? Given we can ignore unknown flags in GO, adding new flags shouldn't be a problem for BC.

I think the point of using file is to avoid using command args. it is not about how do we pass those command args (using a file or specify them in the run command). because command args is defined in the instance and they are compiled into the final executable with the user code. once it is compiled, it is compiled. you can't change it. it is different from python and java.

> I think there is a config you can set to to ignore unknown flags. Please reference https://godoc.org/github.com/jessevdk/go-flags.

we can even write our own flags library to handle that if we really want to. The point is - is there a standard on how do different languages parse and treat *unknown flags* in the same way? If there is a standard, we can adopt it and make all languages use the standard. If that is language specific or library-specific, it is a red flag to me, especially in Go. Please remember the instance is compiled with the user function as the final executable. It is much difficult to fix bug or change library in our instance code. 

Instead if we are using Yaml or Json, they are well-known standards and have clear definitions on what are "unknown properties". I would be much comfortable using them in such situation.

> A user can also copy and paste the command line into a text editor and then parse it however he/she would like.

Well. I will say that the debug experience using a config file (configmap in k8) is much much better than  copying command line, paste it to a text editor, reformat it, and parse it. Especially in a containerized environment.

> What about if we want to support another external scheduler like Nomad or Mesos, it has it own way of mapping files into containers. 

That's why we need a clear abstraction of languages and schedulers, as what I have commented as the first comment in the PR. Commandline doesn't provide you the right abstraction. It is messy as different languages have different instance flags and use different parsing libraries. 

If you are using yaml for defining the instance config, you get a clean and unified way for describing what configuration that the instance uses to invoke functions.

> While all of this is not impossible to do, my question is it really worth it?

It is worth for many reasons:

- Better debug experience (as explained above)
- A unified way of defining the configuration that a (language) instance use to invoke functions, within one scheduler. (as explained above)
- How to ship the instance config is left to schedulers (process, thread, or k8s). This creates a better abstraction between languages and schedulers. (aligned with the long term vision)
- Dynamic configuration loading if needed.
- Versioning the instance configs.
- ...

---

Anyway I think some of the discussions should be done in #4176. 

@jerrypeng I would bringing the conversation back to help @wolfstudy move forward with his change. 

In summary, my take is to avoid using command lines in go instance since that piece of code is shipped to be compiled with user function. Once it is shipped, it is hard to change and we have to support it. If so, I would prefer starting with current instance-config approach, since there is only one command-line argument. It is easier to add more if the approach does creating pain points in k8s runtime. But if we are adding all the command line arguments, it becomes harder to remove them.

Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: > Once it is shipped, it is hard to change and we have to support it. 

I am ok with passing the GO instance the contents of the instance config.  I am just not convinced to method we pass those contents should be a file.  We can also pass the contents of the instance config as command line arguments.   Why can't we just do that? Whether we ignore fields in a yaml file or key-value command line arguments, its kind of the same for purposes of BC.  However if we can just start the GO instance just with command line arguments and have no config file dependency, it will already work with K8s runtime.

If we eventually decide to use the config file approach after discussing in #4176, then lets refactor all the runtimes to support this and deprecate using command line arguments.  I don't see a point to doing this one offshoot  of supporting passing in a config file to an instance just for GO right now.

Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: >  We can also pass the contents of the instance config as command line arguments. 

Let me clarify what do you mean here -

You mean adding a command line argument `--instance-config <config-content>` to go instance. so go instance can use the content passed in to `--instance-config`. 

If that's the case, I am +1 to it. @wolfstudy what do you think?

> If we eventually decide to use the config file approach after discussing in #4176, then lets refactor all the runtimes to support this and deprecate using command line arguments.

+1

Comment (wolfstudy) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: yes, agree, and now we implemented it is like this, after that we can discuss the specific plan in detail in #4176, how about?

Comment (jerrypeng) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: @sijie @wolfstudy yup you can pass in the instance config ```--instance-config <config-content>``` as a json string or as individual arguments:

```
--instanceId <instanceId> --functionDetails <functionDetails> ...
```

either way is fine with me.

Comment (sijie) [reply pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: >  or as individual arguments:

I think my argument above is to avoid using individual arguments since it doesn't work well with BC. so I would prefer `--instance-config <config-content>` to have a fixed number of command args.

**Comment by wolfstudy** [comment]
@sijie @jerrypeng PTAL again

**Comment by wolfstudy** [comment]
run java8 tests

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 20]
```diff
@@ -82,38 +85,174 @@
...
+        }
+
+        if (instanceConfig.getFunctionDetails().getRetryDetails().getMaxMessageRetries() != 0) {
+            goInstanceConfig.setMaxMessageRetries(instanceConfig.getFunctionDetails().getRetryDetails().getMaxMessageRetries());
+        }
+
+        goInstanceConfig.setKillAfterIdleMs(0);
+
+        // Parse the contents of goInstanceConfig into json form string
+        ObjectMapper objectMapper = new ObjectMapper(new JsonFactory());
```
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: please use ObjectMapperFactory.getThreadLocal() instead of creating a new ObjectMapper

File: pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/DefaultSecretsProviderConfigurator.java (hunk 10 above)
Comment (sijie) [reply pulsar-functions/secrets/src/main/java/org/apache/pulsar/functions/secretsproviderconfigurator/DefaultSecretsProviderConfigurator.java]: throw new UnsupportedOperationException();

**Comment by sijie** [comment]
run java8 tests

File: pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java [hunk 21]
```diff
@@ -82,38 +85,174 @@
...
         final List<String> args = new LinkedList<>();
-        if (instanceConfig.getFunctionDetails().getRuntime() ==  Function.FunctionDetails.Runtime.JAVA) {
+        if (instanceConfig.getFunctionDetails().getRuntime() == Function.FunctionDetails.Runtime.JAVA) {
             //no-op
         } else if (instanceConfig.getFunctionDetails().getRuntime() == Function.FunctionDetails.Runtime.PYTHON) {
             // add `extraDependenciesDir` to python package searching path
             if (StringUtils.isNotEmpty(extraDependenciesDir)) {
                 args.add("PYTHONPATH=${PYTHONPATH}:" + extraDependenciesDir);
             }
+        } else if (instanceConfig.getFunctionDetails().getRuntime() == Function.FunctionDetails.Runtime.GO) {
```
Comment (jerrypeng) [inline pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeUtils.java]: I don't think this check is necessary

**Comment by jerrypeng** [comment]
LGTM @srkukarni you want to take a look as well?

**Comment by wolfstudy** [comment]
run java8 tests

**Comment by wolfstudy** [comment]
run java8 tests

**Comment by wolfstudy** [comment]
ping @srkukarni 
do you want to take a look as well?
